### PR TITLE
Update style guide

### PR DESCRIPTION
**Pull request by wemeetagain**
Add more rules to eslint
tsconfig strict==true
Tweaks to CONTRIBUTING

Before fixing linter errors, lets make sure we're not missing any.

### CODE REVIEW

File: .eslintrc
Comment (mpetrunic) [inline .eslintrc]: does that mean it doesn't require sroted imports? 

I kinda like it makes resolving conflicts easier

File: .eslintrc
Comment (mpetrunic) [inline .eslintrc]: isn't this enforcing constants (CONSTANT_VAR) as well?

**Review by mpetrunic — CHANGES_REQUESTED** [review]
Can't find a rule to enforce function param type.

Some rules that I like:
- https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/member-ordering.md
- https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-explicit-any.md - use unknown instead of any if you don't know type, troublesome with ssz (we need some generics solution there)
- https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-unused-vars.md
- prefer-const
- no-var

File: .eslintrc
Comment (wemeetagain) [reply .eslintrc]: I don't think its enforcing constants, but it does allow constants

Comment (mpetrunic) [reply .eslintrc]: Hmm , does it work then? because I saw there is typescript versiomn of camelcase and in recommended settings they disable this rule and enable typescript version of it

File: .eslintrc
Comment (wemeetagain) [reply .eslintrc]: yeah, I think it applies to imports like `import {b, a} from "foo";`

Are you thinking we should enforce sorted keys?

Comment (mpetrunic) [reply .eslintrc]: why false?

**Comment by wemeetagain** [comment]
> Can't find a rule to enforce function param type.

I think `strict: true` in tsconfig will cover this?

>     * https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-explicit-any.md - use unknown instead of any if you don't know type, troublesome with ssz (we need some generics solution there)

I think `strict: true` covers this, but we can add it as well. t`There will definitely need to be some ts-ignore/eslint-disables happening in ssz, currently. (PS This PR should happen after we merge in ssz)

>     * https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/member-ordering.md
>     * https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-unused-vars.md
>     * prefer-const
>     * no-var

:+1:

File: .eslintrc
Comment (wemeetagain) [reply .eslintrc]: I think this was false originally for ssz typescript types, where sometimes it makes sense to have the order different.

**Comment by mpetrunic** [comment]
> I think strict: true in tsconfig will cover this?

nope, I don't think we should use it
https://eslint.org/docs/rules/strict

We need some typescript specfic rule/plugin as normal rules doesn't knwo about typescript types

File: .eslintrc
Comment (mpetrunic) [reply .eslintrc]: how can it affect logic? :joy:

Comment (wemeetagain) [reply .eslintrc]: it doesn't effect logic, it was more of an aesthetic consideration:
eg:
```typescript
const state: BeaconState = {
  // using the order specified in the spec
  genesisTime,
  slot,
  fork,
  ...
  previousJustifiedCheckpoint,
  currentJustifiedCheckpoint,
  finalizedCheckpoint,
};
```
vs
```typescript
const state: BeaconState = {
  balances,
  blockRoots,
  ...
  startShard,
  stateRoots,
};
```

Comment (mpetrunic) [reply .eslintrc]: can we enforce it just on imports?

Comment (wemeetagain) [reply .eslintrc]: I think we can use https://eslint.org/docs/rules/sort-imports

**Review by GregTheGreek — DISMISSED** [review]

**Review by mpetrunic — DISMISSED** [review]

**Review by GregTheGreek — DISMISSED** [review]

**Comment by GregTheGreek** [comment]
this has a conflict probably best hold off anyway

**Comment by mpetrunic** [comment]
@GregTheGreek @wemeetagain I manage to bring this PR up-to-date. I'm up for merging this and make everyone keep their PR clean and aligned with style guide. As it's too hard to keep updating this branch

**Review by wemeetagain — COMMENTED** [review]
LGTM, I can't approve since I opened the PR originally

**Review by mpetrunic — APPROVED** [review]

**Review by GregTheGreek — APPROVED** [review]

**Pull request merged**
